feat: Add `PreviewPhrasePopup`. (M2-7164)

JustinNusca opened the pull request:
### 📝 Description

🔗 [M2-7164](https://mindlogger.atlassian.net/browse/M2-7164): [R2] [Phrase Builder Item Type] Preview a phrase

This PR adds a new popup — `PreviewPhrasePopup`, which can be used to view an approximation of how a specific phrase within a Phrase Builder item may appear to a user within the web app.

This new popup has been connected to "Preview Phrase" button, which was previously non-functional.

### 📸 Screenshots

![2024-10-24 11 42 11](https://github.com/user-attachments/assets/8f1ffd9d-6a16-49c6-acb2-a848f06eeebe)

### 🪤 Peer Testing

While editing an activity within an applet, that has multiple items compatible with the phrase builder item type…

1. Create a new item, and select the type "Phrase Builder".
2. Edit the new phrase to include variations of possible fields, including each of the compatible item types.
3. Press "Preview Phrase" and observe that the popup displays the preview correctly.

[M2-7164]: https://mindlogger.atlassian.net/browse/M2-7164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JustinNusca committed:
Add `PreviewPhrasePopup`.

farmerpaul commented on src/modules/Builder/components/PreviewPhrasePopup/PreviewPhrasePopup.styles.ts:
```diff
@@ -0,0 +1,23 @@
+  height: 56,
+  objectFit: 'cover',
+  overflow: 'hidden',
+  width: 56,
```
Just to stick with our existing dimension units convention (for what it's worth):
```suggestion
  height: '5.6rem',
  objectFit: 'cover',
  overflow: 'hidden',
  width: '5.6rem',
```

JustinNusca replied:
[Updated](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1958/commits/ca4cb25aa34101f132123d46b25ef328dd70354e)!

farmerpaul commented on src/modules/Builder/components/PreviewPhrasePopup/PreviewPhrasePopup.styles.ts:
```diff
@@ -0,0 +1,23 @@
+  borderRadius: variables.borderRadius.lg2,
+  display: 'flex',
+  gap: theme.spacing(2.4),
+  maxHeight: 400,
```
```suggestion
  maxHeight: '40rem',
```

JustinNusca replied:
[Updated](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1958/commits/ca4cb25aa34101f132123d46b25ef328dd70354e)!

farmerpaul commented on src/modules/Builder/features/ActivityItems/ItemConfiguration/InputTypeItems/PhrasalTemplate/PhrasalTemplatePhrase/PhrasalTemplatePhrase.tsx:
```diff
@@ -139,7 +141,7 @@ export const PhrasalTemplatePhrase = ({
               sx={{ gap: 0.8, width: 'max-content' }}
               variant="text"
-              disabled={formState.isDirty || formState.isSubmitting || !formState.isValid}
+              disabled={formState.isSubmitting || !formState.isValid}
```
This button's disabled state is not obvious. The only difference is it lacks a hover effect, and the click action bubbles up to its container. When I first tested your work, I kept thinking my app hadn't refreshed from the right git branch, because clicking it was just toggling the card's collapse state. It wasn't until I realized that the form state was invalid (which has no visual indicator until you've tried to click Save & Publish at least once) that the button was simply disabled.

![CleanShot 2024-10-25 at 10 00 07](https://github.com/user-attachments/assets/c6f3c236-d2c9-493d-96fd-2c55be7a5b33)

So I guess at a minimum, this button variant should have [proper disabled styling](https://www.figma.com/design/LXl9BAJSwtgWuzkxqOqvX8/ML1-Design-System?node-id=57868-142695&m=dev).

Besides the styling issue, I'm also not sure about the UX of clicking a disabled button resulting in its ancestor's click handler to execute. I think it would feel more natural if nothing happened when I clicked it when disabled, and the card expand toggle handler was limited to its surrounding area instead, but I'm not strongly attached to that.

JustinNusca replied:
I pushed up [some changes](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1958/commits/ca4cb25aa34101f132123d46b25ef328dd70354e) that should address this.

First, I fixed the appearance of the disabled state for the button's `text` variant — the theme configuration included some selector specificity conflicts that prevented the correct text color from being applied.

Since this is a global change, I clicked around other locations where we were using the `text` variant with a (conditionally) disabled state (eg, The "Feedback" & "Download Latest Report" buttons in `/dashboard/{appletId}/participants/{subjectId}/activities/{activityId}/{summary | responses}`) — there's not that many and there doesn't seem to be any adverse effect elsewhere 😀.

So that solves one problem. The second issue — the nested interactive elements behaviour — is almost identical in cause to the one I nagged you about in your [PDP pull request](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1943). 😭

My solution here is a bit of a kludge. I was hoping to do something similar to yours in that PR, but we're using slightly different components (`Accordion` vs `Collapse`) and there's a slightly different desired behaviour (we want a `disabled` button to continue to prevent clicks on the parent element). 

In the end I just added an additional `onClick` with a comment explaining the purpose on the wrapper node around both this and the "Delete Phrase" button, that simply calls `stopPropagation`. This works to stand-in as a handler for the button when it's in a disabled state, and avoids misclicks if you happen to press in the ambiguous boundary between the two buttons.

farmerpaul commented on src/modules/Builder/components/PreviewPhrasePopup/PreviewPhrasePopup.tsx:
```diff
@@ -0,0 +1,72 @@
+        case 'sentence':
+          return <Box component="span">{fieldValue.text}</Box>;
+        case 'line_break':
+          return <Box component="hr" sx={{ m: 0, height: 32, border: 'none' }} />;
```
Line breaks don't actually insert a gap… this is what the web app adds when it encounters one:
```suggestion
          return <br />;
```

JustinNusca replied:
Hmm, are you sure there should be no appearance of a gap? I couldn't find explicit direction about the preview card, but I was following the [spec for the web app, which specifies that line breaks should insert a 32px gap](https://www.figma.com/design/3wcZ6bxK2QVo733ADxVjBw/SSI-2024?node-id=692-76427&t=BLw76cfZQzfaJAK6-4). 

It doesn't seem like there's a conventional/cross-browser way to style a `<br>` tag like this.

farmerpaul replied:
Ah, my bad. I didn't realize this was a [bug in the web app code](https://github.com/ChildMindInstitute/mindlogger-web-refactor/blob/4b938c806d4d5dc356998dd944b845c5bb8300f5/src/entities/activity/ui/items/ActionPlan/Phrase.tsx#L46). I was using the web app as the source of truth rather than Figma. I guess I'll set a reminder to create a ticket for this next week, unless you have time to create a quick PR for the web app… 😏

JustinNusca replied:
I was somewhat reluctant to open a new PR to address this, given all of @qiushihe's work that's going into refactoring the code in that exact area, but I left a [comment on the PR](https://github.com/ChildMindInstitute/mindlogger-web-refactor/pull/541/files#r1817086469) that should hopefully help explain and allow her to take care of it 🤞.

farmerpaul reviewed (approved):
Awesome work again @JustinNusca, so quick! I just pointed out a couple minor things, and one UX issue, but it's easy enough to fix that I'm confident to pre-approve. ✨ 🚀 ✨

JustinNusca committed:
Address PR comments.

- Use explicit rem units in styling.
- Update Button `text` variant disabled state.
